include/cell/copy/global_to_shared.hpp

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+#include "cell/copy/constants.hpp"
+#include "cell/copy/dyn_copy.hpp"
+#include "cell/copy/warp.hpp"
+#include "cell/traits/base.hpp"
+#include "types/mod.hpp"
+
+namespace tiledcuda::cell::copy {
+
+using namespace traits;
+namespace tl = tile_layout;
+
+template <typename Global_, typename Shared_, const int kThreads_,
+          const tl::Layout kType, typename Base>
+struct GlobalToSharedLoaderImpl {
+    using Global = Global_;
+    using Shared = Shared_;
+    using DType = typename Global::DType;
+
+    DEVICE void operator()(const DType* src, DType* dst);
+};
+
+template <typename Global_, typename Shared_, const int kThreads_,
+          typename Base>
+struct GlobalToSharedLoaderImpl<Global_, Shared_, kThreads_,
+                                tl::Layout::kSwizzledRowMajor, Base> {
+    using Global = Global_;
+    using Shared = Shared_;
+    using DType = typename Global::DType;
+
+    static constexpr int kThreads = kThreads_;
+
+    static constexpr int kRows = Global::kRows;
+    static constexpr int kCols = Global::kCols;
+    static constexpr int kShmRows = Shared::kRows;
+    static constexpr int kShmCols = Shared::kCols;
+
+    // To avoid bank conflict, the shared memory requires a swizzled layout
+    static const int kSwizzleMode = kShmCols % 32 ? 1 : 0;
+    using Swizzled =
+        tl::SwizzledRowMajor<DType, kShmRows, kShmCols, kSwizzleMode>;
+
+    using SrcLayout = tl::RowMajor<kRows, kCols, kCols>;
+    using DstLayout = typename Swizzled::SmemLayout;
+
+    // threads in a thread block are laid out as a 2D tile
+    // that has a shape of kThreadsRows x kThreadsCols.
+    static constexpr int kThreadsCols = kShmCols / Base::kNumPerAccess;
+    static constexpr int kThreadsRows = kThreads / kThreadsCols;
+
+    using ThreadLayout = tl::RowMajor<kThreadsRows, kThreadsCols, kThreadsCols>;
+
+    using ValueLayout = Layout<Shape<_1, Int<Base::kNumPerAccess>>>;
+
+#ifdef CP_ASYNC_SM80_ENABLED
+    using CopyInst =
+        Copy_Atom<SM80_CP_ASYNC_CACHEGLOBAL<cute::uint128_t>, DType>;
+#else
+    using CopyInst = Copy_Atom<DefaultCopy, DType>;
+#endif
+
+    using TiledCopy =
+        decltype(make_tiled_copy(CopyInst{}, ThreadLayout{}, ValueLayout{}));
+
+    DEVICE void operator()(const DType* src, DType* dst) {
+        copy_2d_tile_g2s(src, dst, SrcLayout{}, DstLayout{}, TiledCopy{});
+    }
+};
+
+template <typename Global_, typename Shared_, const int kThreads_,
+          const tl::Layout kType, typename Base>
+struct SharedToGlobalStorerImpl {
+    using Global = Global_;
+    using Shared = Shared_;
+    using DType = typename Global::DType;
+
+    DEVICE void operator()(const DType* src, DType* dst);
+};
+
+template <typename Global_, typename Shared_, const int kThreads_,
+          typename Base>
+struct SharedToGlobalStorerImpl<Global_, Shared_, kThreads_,
```

Comment:
I suggest hiding `Base` from the template parameters. Since `Base` only depends on `DType` to calculate some hardware-related parameters, and `SharedToGlobalLoader/Storer` already holds `DType`, exposing `Base` in the template parameters seems unnecessary.

And, I would like to know your considerations on exposing this.

Reply:
Your perspective is better, I did not consider these issues carefully enough.